File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/person/AgeAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.person;

import java.util.List;

import com.truveta.opentoken.attributes.BaseAttribute;
import com.truveta.opentoken.attributes.validation.AgeRangeValidator;
import com.truveta.opentoken.attributes.validation.RegexValidator;

/**
 * Represents the age attribute.
 * 
 * This class extends BaseAttribute and provides functionality for working with
 * age fields. It recognizes "Age" as the primary alias for this attribute type.
 * 
 * The attribute performs normalization on input values, trimming whitespace and
 * ensuring the value is a valid integer.
 * 
 * The attribute also performs validation on input values, ensuring they:
 * - Are numeric (integers only)
 * - Fall within an acceptable range (0-120)
 */
public class AgeAttribute extends BaseAttribute {

    private static final String NAME = "Age";
    private static final String[] ALIASES = new String[] { NAME };

    /**
     * Regular expression pattern for validating age format.
     * 
     * This regex validates that the input is:
     * - A positive integer (no decimals)
     * 
     * Note: Leading/trailing whitespace is allowed by the regex,
     * but is removed during normalization, not validation.
     */
    private static final String AGE_REGEX = "^\\s*\\d+\\s*$";

    public AgeAttribute() {
        super(List.of(
                new RegexValidator(AGE_REGEX),
                new AgeRangeValidator()));
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String[] getAliases() {
        return ALIASES;
    }

    @Override
    public String normalize(String value) {
        if (value == null) {
            throw new IllegalArgumentException("Age value cannot be null");
        }

        // Trim whitespace and validate it's a number
        String trimmed = value.trim();

        try {
            int age = Integer.parseInt(trimmed);
            // Return the trimmed integer as a string
            return String.valueOf(age);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Invalid age format: " + value);
        }
    }

}

File: lib/python/opentoken/src/main/opentoken/attributes/person/age_attribute.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import List
import re
from opentoken.attributes.base_attribute import BaseAttribute
from opentoken.attributes.validation import RegexValidator
from opentoken.attributes.validation.age_range_validator import AgeRangeValidator


class AgeAttribute(BaseAttribute):
    """Represents the age attribute.

    This class extends BaseAttribute and provides functionality for working with
    age fields. It recognizes "Age" as a valid alias for this attribute type.

    The attribute performs normalization on input values by trimming whitespace
    and validates that the age is a valid integer between 0 and 120.
    """

    NAME = "Age"
    ALIASES = [NAME]

    # Regular expression pattern for validating integer format with optional whitespace
    VALIDATION_PATTERN = re.compile(r"^\s*\d+\s*$")

    def __init__(self):
        validation_rules = [
            RegexValidator(self.VALIDATION_PATTERN),
            AgeRangeValidator()
        ]
        super().__init__(validation_rules)

    def get_name(self) -> str:
        """Get the name of the attribute.

        Returns:
            str: The name "Age"
        """
        return self.NAME

    def get_aliases(self) -> List[str]:
        """Get the aliases for the attribute.

        Returns:
            List[str]: A list containing the aliases for this attribute
        """
        return self.ALIASES.copy()

    def normalize(self, value: str) -> str:
        """Normalize the age value by trimming whitespace.

        Args:
            value: The age string to normalize

        Returns:
            str: The trimmed age value

        Raises:
            ValueError: If the age is not a valid integer
        """
        trimmed = value.strip()
        
        # Validate it's a valid integer
        try:
            int(trimmed)
            return trimmed
        except ValueError:
            raise ValueError(f"Invalid age format: {value}")
